localstack/aws/handlers/cors.py

```diff
@@ -145,19 +145,19 @@ def should_enforce_self_managed_service(context: RequestContext) -> bool:
                     the targeting service
     :return: True if the CORS rules should be enforced in here.
     """
-    if config.DISABLE_CUSTOM_CORS_S3 and config.DISABLE_CUSTOM_CORS_APIGATEWAY:
-        return True
     # allow only certain api calls without checking origin
-    if context.service:
-        service_name = context.service.service_name
-        if not config.DISABLE_CUSTOM_CORS_S3 and service_name == "s3":
+    if not config.DISABLE_CUSTOM_CORS_S3:
+        if context.service and context.service.service_name == "s3":
             return False
-        if not config.DISABLE_CUSTOM_CORS_APIGATEWAY and service_name == "apigateway":
-            is_user_request = (
-                PATH_USER_REQUEST in context.request.path or ".execute-api." in context.request.host
-            )
-            if is_user_request:
-                return False
+
+    if not config.DISABLE_CUSTOM_CORS_APIGATEWAY:
+        # we don't check for service_name == "apigw" here because ``.execute-api.`` can be either apigw v1 or v2
+        is_user_request = (
+            ".execute-api." in context.request.host or PATH_USER_REQUEST in context.request.path
```

Comment:
I think there is a very small edge case that could actually slip through this here, which would mean that a request which should potentially be blocked is not checked.
Ultimately, I think this if has to be equal to the check in the service router:
https://github.com/localstack/localstack/blob/9af4246ef854ab27a2f420799ce45caedf8d2f26/localstack/aws/protocol/service_router.py#L181-L183

Actually, we have multiple rules in the service router for the `.execute-api.` in the host, so I think the only difference would be the surrounding slashes in the path for `PATH_USER_REQUEST`:
https://github.com/localstack/localstack/blob/451968fabb223e189323df02a8d08fac110fddec/localstack/aws/protocol/service_router.py#L145-L147

Which means the following check should be equivalent to the service router rules, ultimately closing this small gap:
```suggestion
            ".execute-api." in context.request.host or f"/{PATH_USER_REQUEST}/" in context.request.path
```

What do you think?

Reply:
Do you think that's blocking? To me the check in the service router is only stricter. Which means the CORS check would not let something pass with this fix that it did not let before.

Reply:
Also, I think this legacy condition: `host.endswith(LOCALHOST_HOSTNAME) and "execute-api" in host` is not really good anymore with the changes from Simon what's after the prefix is not matched anymore. I think this code "should" be dead code.

To resume, we do not need to have the service name set for the router, and the APIGW routes will match with or without the service set in the context.

The issue created by the previous PR and now fixed with this PR is that we would set the service early for a request that what sent with the SDK/AWS-CLI (so it signed the request, the service name parser would return very early with the signing name), but the `execute-api` endpoint. See https://docs.aws.amazon.com/cli/latest/reference/apigatewaymanagementapi/post-to-connection.html#examples
And because the service name parser returns early if the service is already set, we would skip and keep `apigateway` (v1) set for that request. 

So in the end, the fact that we removed it from the CORS check means we should be able to remove this legacy condition in the router too.

Reply:
I see what you mean, thanks for clarifying.
I think both previous and current approach are very approximative, and do not really guarantee a correct detection of APIGW. Even if `PATH_USER_REQUEST`  is between `/`, it does not mean it's going to match with a route, the first one linked here does not have a mandatory slash on the right. Also, the `PATH_USER_REQUEST` between slashes could be anywhere in the path, and the route might not match. So it's already a pretty flawed approach, and the change makes it a bit weaker only. 

Also, I think with our current approach of both CORS and CRSF in the same check, we're not really weakening CORS, because we will just not return CORS headers when we should, which in the end doesn't allow the browser to read the response and block it. But the same check is used for CRSF, which means we're not blocking it either.

We could maybe create stricter conditions, being that the current request matches against APIGW registered routes? I can have a quick look into it, to make the condition stricter (regex matching on the host or the path), but that's a bit expensive.

Reply:
I'm currently working on a PR and I will add tests for it. 
I think we will need to revisit anyway, because at some point it would be good to be able to execute those routes from the Web App somehow, but currently you would have to setup CORS manually in your API for that. I'll keep that in mind for the future. 
CORS handling is a bit trickier in APIGW than S3 because there are no explicit APIs to manage it, it's either MOCK responses for `OPTIONS` or your lambda returning it... not an easy one.

This is the PR: https://github.com/localstack/localstack/pull/10688

Thanks a lot for raising this!